smithy-aws-cloudformation-traits/src/main/java/software/amazon/smithy/aws/cloudformation/traits/CfnResourceTraitValidator.java

```diff
@@ -0,0 +1,91 @@
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.aws.cloudformation.traits;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.knowledge.TopDownIndex;
+import software.amazon.smithy.model.selector.PathFinder;
+import software.amazon.smithy.model.selector.Selector;
+import software.amazon.smithy.model.shapes.ResourceShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.validation.AbstractValidator;
+import software.amazon.smithy.model.validation.ValidationEvent;
+import software.amazon.smithy.utils.ListUtils;
+
+/**
+ * Validates that all shapes referenced by the additionalSchemas property of the
+ * {@code @cfnResource} trait are also bound to any service that the resource is
+ * bound to.
+ */
+public final class CfnResourceTraitValidator  extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        if (!model.isTraitApplied(CfnResourceTrait.class)) {
+            return ListUtils.of();
+        }
+
+        List<ValidationEvent> events = new ArrayList<>();
+        TopDownIndex topDownIndex = TopDownIndex.of(model);
+        Set<ServiceShape> serviceShapes = model.getServiceShapes();
+
+        for (ResourceShape resourceShape : model.getResourceShapesWithTrait(CfnResourceTrait.class)) {
+            CfnResourceTrait cfnResourceTrait = resourceShape.expectTrait(CfnResourceTrait.class);
+
+            // Short-circuit if there are no additionalSchemas for this resource.
+            if (cfnResourceTrait.getAdditionalSchemas().isEmpty()) {
+                continue;
+            }
+
+            // Validate that the for each service that contains the resource.
+            for (ServiceShape serviceShape : serviceShapes) {
+                if (!topDownIndex.getContainedResources(serviceShape).contains(resourceShape)) {
+                    continue;
+                }
+
+                events.addAll(validateAdditionalSchemaBindings(model, serviceShape, resourceShape, cfnResourceTrait));
+            }
+        }
+
+        return events;
+    }
+
+    private List<ValidationEvent> validateAdditionalSchemaBindings(
+            Model model,
+            ServiceShape service,
+            ResourceShape resource,
+            CfnResourceTrait trait
+    ) {
+        List<ValidationEvent> events = new ArrayList<>();
+
+        for (ShapeId additionalSchemaId : trait.getAdditionalSchemas()) {
+            Selector selector = Selector.parse("[id = " + additionalSchemaId + "]");
+
+            // Emit an event if we don't have a path to the additionalSchema from the
+            // service shape this resource is bound to.
+            if (PathFinder.create(model).search(service.getId(), selector).isEmpty()) {
```

Comment:
nit: This is very clever, but wouldn't it be more performant to compute the shapes connected to the service once and then do a contains check for each additional schema?